### PR TITLE
Expose `getUnaryFlags` and `getBinaryFlags` as xsmm utils (NFC)

### PR DESCRIPTION
**Pull request by chelini**
Needed in #810 for better XSMM verification.

### CODE REVIEW

File: lib/TPP/Dialect/Xsmm/XsmmUtils.cpp [hunk 1]
```diff
@@ -148,6 +148,134 @@ FailureOr<BinaryInfo> getBinaryInfo(Value lhs, BinaryFlags lhsFlag, Value rhs,
...
+      reshapeOutputShape[i] = lowerRankDim;
+    else if (higherRankDim != lowerRankDim)
+      assert(false && "bCast semantics for identity op broken");
+  }
+}
+
+FailureOr<UnaryFlags> getUnaryFlags(Type inputType, Type outputType) {
+  // There are multiple ways to define a scalar.  f32, memref<1x1xf32> or
+  // memref<f32>. Handle f32, and memref<1x1xf32>. memref<f32> is not allowed
+  // in tpp at the moment.
```
Comment (rengolin) [inline lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:188]: I can't remember if this was a hard restriction of just we didn't care enough. AFAICS, `memref<1xf32>`, `memref<1x1xf32>`, etc are identical to `memref<f32>` in memory representation (not logically). XSMM shouldn't care about the logical semantics, right?

File: lib/TPP/Dialect/Xsmm/XsmmUtils.cpp [hunk 2]
```diff
@@ -148,6 +148,134 @@ FailureOr<BinaryInfo> getBinaryInfo(Value lhs, BinaryFlags lhsFlag, Value rhs,
...
+
+  ArrayRef<int64_t> shapeOutput = outputType.cast<ShapedType>().getShape();
+  assert(shapeOutput.size() >= shapeInput.size() &&
+         "output rank must be >= input rank");
+  SmallVector<int64_t> bShapeInput;
+  computeBcastShapeInput(shapeOutput, shapeInput, bShapeInput);
+  assert(shapeOutput.size() == bShapeInput.size());
+  shapeInput = bShapeInput;
+
+  if (shapeInput[1] == 1 && shapeOutput[1] > 1)
```
Comment (rengolin) [inline lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:204]: Do we need to check if the rank is 2?

Is `2 x 1 x 2 -> 2 x 2 x 2` a row broadcast?

File: lib/TPP/Dialect/Xsmm/XsmmUtils.cpp [hunk 3]
```diff
@@ -148,6 +148,134 @@ FailureOr<BinaryInfo> getBinaryInfo(Value lhs, BinaryFlags lhsFlag, Value rhs,
...
+    return xsmm::UnaryFlags::BCAST_ROW;
+
+  if (shapeInput[0] == 1 && shapeOutput[0] > 1)
+    return xsmm::UnaryFlags::BCAST_COL;
+
+  if (shapeInput[0] == shapeOutput[0] && shapeInput[1] == shapeOutput[1])
```
Comment (rengolin) [inline lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:210]: Can't remember if you can do `(shapeInput == shapeOutput)`.

File: lib/TPP/Dialect/Xsmm/XsmmUtils.cpp [hunk 4]
```diff
@@ -148,6 +148,134 @@ FailureOr<BinaryInfo> getBinaryInfo(Value lhs, BinaryFlags lhsFlag, Value rhs,
...
+    return xsmm::UnaryFlags::NONE;
+  return failure();
+}
+
+FailureOr<BinaryFlags> getBinaryFlags(Type operandType, Type outputType,
```
Comment (rengolin) [inline lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:215]: This requires the caller to validate if the "other" operand has the same shape as the output. I'd add all three operands and make that check here (even as a static call on the same unit).

File: lib/TPP/Dialect/Xsmm/XsmmUtils.cpp [hunk 5]
```diff
@@ -148,6 +148,134 @@ FailureOr<BinaryInfo> getBinaryInfo(Value lhs, BinaryFlags lhsFlag, Value rhs,
...
+    assert(false && "unrechable");
+  };
+
+  // Multiple way to define a scalar. Check if the memref
+  // is a scalar here.
+  auto isOne = [](int64_t val) { return val == 1; };
+  if (llvm::all_of(bOperandShape, isOne))
+    return getBCastEnum(BCastType::SCALAR, operandNumber);
+
+  if (bOperandShape[1] == 1 && shapeOutput[1] > 1)
```
Comment (rengolin) [inline lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:270]: Same question about rank as in the unary case.

File: lib/TPP/Dialect/Xsmm/XsmmUtils.cpp (hunk 1 above)
Comment (chelini) [reply lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:188]: We used to pass memref to the runtime and we had to match the memref ABI, I can relax this checks now that we pass pointers.

File: lib/TPP/Dialect/Xsmm/XsmmUtils.cpp (hunk 3 above)
Comment (chelini) [reply lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:210]: Yes, we can.

File: lib/TPP/Dialect/Xsmm/XsmmUtils.cpp (hunk 4 above)
Comment (chelini) [reply lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:215]: We can have different flags per operand as far as I understood from LIBXSMM. Basically, the flags are computed per operand with respect to the output. Does this sound wrong to you?

Comment (rengolin) [reply lib/TPP/Dialect/Xsmm/XsmmUtils.cpp:215]: Good point. In theory we could have `(<10x1xf32>, <20xf32>) -> <10x20xf32>`, which would have both row/col broadcasts on different operands.

**Pull request merged**